src/array.jl

```diff
@@ -389,17 +389,26 @@ end
 # Methods preserving levels and more efficient than AbstractArray fallbacks
 copy(A::CategoricalArray) = deepcopy(A)
 
-function copy!(dest::CategoricalArray{T, N}, dstart::Integer,
-               src::CategoricalArray{T, N}, sstart::Integer,
-               n::Integer=length(src)-sstart+1) where {T, N}
+function copy!(dest::CategoricalArray{<:Union{T, Null}, N},
+               dstart::Integer,
+               src::S,
+               sstart::Integer,
+               n::Integer=length(src)-sstart+1) where
+    {T, N, S <: Union{CategoricalArray{<:Union{T ,Null}, N},
+                      SubArray{<:Any, <:Any, <:CategoricalArray{<:Union{T, Null}, N}}}}
     destinds, srcinds = linearindices(dest), linearindices(src)
     (dstart ∈ destinds && dstart+n-1 ∈ destinds) || throw(BoundsError(dest, dstart:dstart+n-1))
     (sstart ∈ srcinds  && sstart+n-1 ∈ srcinds)  || throw(BoundsError(src,  sstart:sstart+n-1))
     n == 0 && return dest
     n < 0 && throw(ArgumentError(string("tried to copy n=", n, " elements, but n should be nonnegative")))
 
     drefs = dest.refs
-    srefs = src.refs
+    if isa(src, SubArray)
+        srefs = view(src.parent.refs, src.indexes...)
+    else
+        srefs = src.refs
+    end
+    spool = isa(src, SubArray) ? src.parent.pool : src.pool
```

Comment:
That's a good example why it might be good to have `CategoricalArrays.pool(A::CategoricalArray)` and `CategoricalArrays.pool(A::SubArray{CategoricalArray})`; the same for `refs()`.
All type-dependent logic would be automatically handled by dispatch.
It would also be easy to extend to handle e.g. `ReshapedArray{CategoricalArray}`.